README.md
# FFI::TinyCC::Inline [![Build Status](https://secure.travis-ci.org/Perl5-FFI/FFI-TinyCC-Inline.png)](http://travis-ci.org/Perl5-FFI/FFI-TinyCC-Inline)

Embed Tiny C code in your Perl program

# SYNOPSIS

    use FFI::TinyCC::Inline qw( tcc_inline );
    
    tcc_inline q{
      int square(int num)
      {
        return num*num;
      }
    };
    
    print square(4), "\n"; # prints 16

    use FFI::TinyCC::Inline qw( tcc_eval );
    
    # sets value to 6:
    my $value = tcc_eval q{
      int main(int a, int b, int c)
      {
        return a + b + c;
      }
    }, 1, 2, 3;

# DESCRIPTION

This module provides a simplified interface to FFI::TinyCC, that allows you
to write Perl subs in C.  It is inspired by [XS::TCC](https://metacpan.org/pod/XS::TCC), but it uses [FFI::Platypus](https://metacpan.org/pod/FFI::Platypus)
to create bindings instead of XS.

# OPTIONS

You can specify Tiny C options using the scoped pragmata, like so:

    use FFI::TinyCC::Inline options => "-I/foo/include -L/foo/lib -DFOO=1";
    
    # prints 1
    print tcc_eval q{
    #include <foo.h> /* will search /foo/include
    int main()
    {
      return FOO; /* defined and set to 1 */
    }
    };

# FUNCTIONS

## tcc\_inline

    tcc_inline $c_code;

Compile the given C code using Tiny C and inject any functions found into the
current package.  An exception will be thrown if the code fails to compile, or if
[FFI::TinyCC::Inline](https://metacpan.org/pod/FFI::TinyCC::Inline) does not recognize one of the argument or return
types.

    tcc_inline q{
      int foo(int a, int b, int c)
      {
        return a + b + c;
      }
    };
    
    print foo(1,2,3), "\n"; # prints 6

The special argument type of `(int argc, char **argv)` is recognized and
will be translated from the list of arguments passed in.  Example:

    tcc_inline q{
      void foo(int argc, const char **argv)
      {
        int i;
        for(i=0; i<argc; i++)
        {
          puts(argv[i]);
        } 
      }
    };
    
    foo("one", "two", "three"); # prints "one\ntwo\nthree\n"

## tcc\_eval

    tcc_eval $c_code, @arguments;

This compiles the C code and executes the `main` function, passing in the given arguments.
Returns the result.

# SEE ALSO

- [FFI::TinyCC](https://metacpan.org/pod/FFI::TinyCC)
- [C::Blocks](https://metacpan.org/pod/C::Blocks)

# BUNDLED SOFTWARE

This package also comes with a parser that was shamelessly stolen from [XS::TCC](https://metacpan.org/pod/XS::TCC),
which I strongly suspect was itself shamelessly "borrowed" from 
[Inline::C::Parser::RegExp](https://metacpan.org/pod/Inline::C::Parser::RegExp)

The license details for the parser are:

Copyright 2002 Brian Ingerson
Copyright 2008, 2010-2012 Sisyphus
Copyright 2013 Steffen Muellero

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

# AUTHOR

Author: Graham Ollis <plicease@cpan.org>

Contributors:

aero

Dylan Cali (calid)

pipcet

# COPYRIGHT AND LICENSE

This software is copyright (c) 2015-2018 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.28      2018-06-07 07:46:56 -0400
  - Production release identical to 0.27_01

0.27_01   2018-06-06 14:40:32 -0400
  - FFI::TinyCC and FFI::TinyCC::Inline are now separately distributed.
    Older entries include the history of BOTH modules.
  - Perl 5.10 or later is now required.

0.26      2017-07-31 17:36:33 -0400
  - Remove dependency on IPC::System::Simple

0.25      2017-07-31 17:10:55 -0400
  - Migrated to internally using File::ShareDir::Dist instead of File::ShareDir
  - Migrated to internally using Path::Tiny instead of Path::Class

0.24      2017-01-31 00:04:51 -0500
  - Mark executable output as possibly not supported.  Anywhere.
  - Remove deprecated support for old FFI library

0.23      2017-01-06 15:12:55 -0500
  - Updated documentation to reflect FFI::Platypus preference for
    not using Declare interface

0.22      2015-08-06 09:37:59 -0400
  - Refinement of detect_sysinclude_path
  - libtcc.so is built during the build stage instead of the configure
    stage during dist install
  - Numerous small but important documentation corrections
  - Add link to #native irc channel in dist meta data
    (join us for Alien and FFI discussions)

0.21      2015-03-11 22:24:12 -0400
  - Fix bug where get_symbol method was not correctly throwing an exception
    on error (gh#4 pipcet++

0.20      2015-03-11 12:57:12 -0400
  - Fix bug in detect_sysinclude_path on Perl 5.18 and earlier introduced
    in 0.19

0.19      2015-03-11 09:19:47 -0400
  - detect_sysinclude_path returns the directories added to the system
    include path (gh#3 suggested by calid++)

0.18      2015-03-11 06:44:56 -0400
  - Production release identical to 0.17_02

0.17_02   2015-03-10 09:17:52 -0400
  - Removed unintended linux system include path from detect_sysinclude_path.t

0.17_01   2015-03-10 05:48:35 -0400
  - Added method detect_sysinclude_path, which attempts to find the correct
    sysinclude path for you (so that TinyCC can find things like <stdio.h>)

0.17      2015-01-27 21:47:45 -0500
  - Update FFI::Platypus syntax in examples and documentation
    A couple of places were using conventions that were removed
    from Platypus before it went to CPAN.

0.16      2015-01-25 03:58:01 -0500
  - FFI::TinyCC::Inline options is not supported on Perl 5.8
    a useful diagnostic will be printed if you try to use it on 5.8
    tests that test options will skip on 5.8

0.15      2015-01-23 17:03:42 -0500
  - use FFI::Platypus::Type::StringArray instead of reinventing it ourselves

0.14      2015-01-22 06:38:34 -0500
  - CPAN release identical to 0.12

0.12      2015-01-19 03:15:27 -0500
  - FFI::Platypus 0.09 compatability
  - Development release not posted on CPAN

0.11      2015-01-15 18:51:26 -0500
  - migrate to FFI::Platypus.  Should still interact with old FFI library
    as before.
  - Your old FFI library code may break if you are not explicitly using
    that old FFI library, and depending on FFI::TinyCC to pull it in for you.
    Stop doing that.
  - Development release not posted on CPAN

0.10      2015-01-08 10:09:47 -0500
  - get_ffi_raw method is deprecated and issue a warning.
  - Old FFI library will be removed as a prerequisite in the near future
    if you are using both the old FFI library and FFI::TinyCC then you need
    to explicitly declare BOTH as prerequisites.

0.09      2014-11-17 06:31:51 -0500
  - Skip unsupported tests on arm linux

0.08      2014-11-09 07:47:07 -0500
  - add Debian kFreeBSD support (skip some unsupported tests)

0.07      2014-11-05 20:53:30 -0500
  - add OS X support
    remove trailing null characters from libtcc.a members (possible Archive::Ar bug)
    skip dll and exe tests for features not supported on OS X

0.06      2014-11-05 18:07:06 -0500
  - skip t/output_file_exe.t test on *BSD and darwin
    that doesn't seem to be supported there at the moment

0.05      2014-11-05 07:22:26 -0500
  - fixed Windows breakage due to hard coded path (doh!)

0.04      2014-11-04 16:58:12 -0500
  - better test diagnostics

0.03      2014-11-04 05:31:55 -0500
  - testing improvements

0.02      2014-10-28 15:16:44 -0400
  - added tcc_eval and tcc_inline interfaces in FFI::TinyCC::Inline

0.01      2014-10-28 09:29:55 -0400
  - initial version


lib/FFI/TinyCC/Inline.pm
package FFI::TinyCC::Inline;

use strict;
use warnings;
use 5.010;
use FFI::Platypus;
use FFI::TinyCC;
use Carp qw( croak );
use base qw( Exporter );

our @EXPORT_OK = qw( tcc_inline tcc_eval );
our @EXPORT = @EXPORT_OK;

# ABSTRACT: Embed Tiny C code in your Perl program
# VERSION

=head1 SYNOPSIS

 use FFI::TinyCC::Inline qw( tcc_inline );
 
 tcc_inline q{
   int square(int num)
   {
     return num*num;
   }
 };
 
 print square(4), "\n"; # prints 16

 use FFI::TinyCC::Inline qw( tcc_eval );
 
 # sets value to 6:
 my $value = tcc_eval q{
   int main(int a, int b, int c)
   {
     return a + b + c;
   }
 }, 1, 2, 3;

=head1 DESCRIPTION

This module provides a simplified interface to FFI::TinyCC, that allows you
to write Perl subs in C.  It is inspired by L<XS::TCC>, but it uses L<FFI::Platypus>
to create bindings instead of XS.

=cut

my $ffi = FFI::Platypus->new;
$ffi->load_custom_type( 'StringArray' => 'string_array' );

# TODO: support platypus types like pointers and arrays
my %typemap = (
  'int'            => 'int',
  'signed int'     => 'signed int',
  'unsigned int'   => 'unsigned int',
  'void'           => 'void',
  'short'          => 'short',
  'signed short'   => 'signed short',
  'unsigned short' => 'unsigned short',
  'long'           => 'long',
  'signed long'    => 'signed long',
  'unsigned long'  => 'unsigned long',
  'char'           => 'char',
  'signed char'    => 'signed char',
  'unsigned char'  => 'unsigned char',
  'float'          => 'float',
  'double'         => 'double',
  'char *'         => 'string',
);

sub _typemap ($)
{
  my($type) = @_;
  $type =~ s{^const }{};
  return $typemap{$type}
    if defined $typemap{$type};
  return 'opaque' if $type =~ /\*$/;
  croak "unknown type: $type";
}

sub _generate_sub ($$$)
{
  my($func_name, $func, $tcc) = @_;
  my $sub;
  
  my $address = $tcc->get_symbol($func_name);
  
  if(@{ $func->{arg_types} } == 2
  && $func->{arg_types}->[0] eq 'int'
  && $func->{arg_types}->[1] =~ /^(const |)char \*\*$/)
  {
    my $f = $ffi->function($address => ['int','string_array'] => _typemap $func->{return_type});
    $sub = sub {
      $f->call(scalar @_, \@_);
    };
  }
  else
  {
    my $f = $ffi->function($address => [map { _typemap $_ } @{ $func->{arg_types} }] => _typemap $func->{return_type});
    $sub = sub { $f->call(@_) };
  }
  
  $sub;
}

=head1 OPTIONS

You can specify Tiny C options using the scoped pragmata, like so:

 use FFI::TinyCC::Inline options => "-I/foo/include -L/foo/lib -DFOO=1";
 
 # prints 1
 print tcc_eval q{
 #include <foo.h> /* will search /foo/include
 int main()
 {
   return FOO; /* defined and set to 1 */
 }
 };

=cut

sub import
{
  my($class, @rest) = @_;
  
  if(defined $rest[0] && defined $rest[1]
  && $rest[0] eq 'options')
  {
    if($] >= 5.010)
    {
      shift @rest;
      $^H{"FFI::TinyCC::Inline/options"} = shift @rest;
    }
    else
    {
      croak "options not supported on Perl 5.8";
    }
  }
  
  return unless @rest > 0;

  @_ = ($class, @rest);
  goto &Exporter::import;
}

=head1 FUNCTIONS

=head2 tcc_inline

 tcc_inline $c_code;

Compile the given C code using Tiny C and inject any functions found into the
current package.  An exception will be thrown if the code fails to compile, or if
L<FFI::TinyCC::Inline> does not recognize one of the argument or return
types.

 tcc_inline q{
   int foo(int a, int b, int c)
   {
     return a + b + c;
   }
 };
 
 print foo(1,2,3), "\n"; # prints 6

The special argument type of C<(int argc, char **argv)> is recognized and
will be translated from the list of arguments passed in.  Example:

 tcc_inline q{
   void foo(int argc, const char **argv)
   {
     int i;
     for(i=0; i<argc; i++)
     {
       puts(argv[i]);
     } 
   }
 };
 
 foo("one", "two", "three"); # prints "one\ntwo\nthree\n"

=cut

sub tcc_inline ($)
{
  my($code) = @_;
  my $caller = caller;
  
  my $tcc = FFI::TinyCC->new(_no_free_store => 1);
  
  my $h = (caller(0))[10];
  if($h->{"FFI::TinyCC::Inline/options"})
  { $tcc->set_options($h->{"FFI::TinyCC::Inline/options"}) }

  $tcc->compile_string($code);
  my $meta = FFI::TinyCC::Parser->extract_function_metadata($code);
  foreach my $func_name (keys %{ $meta->{functions} })
  {
    my $sub = _generate_sub($func_name, $meta->{functions}->{$func_name}, $tcc);
    no strict 'refs';
    *{join '::', $caller, $func_name} = $sub;
  }
  ();
}

=head2 tcc_eval

 tcc_eval $c_code, @arguments;

This compiles the C code and executes the C<main> function, passing in the given arguments.
Returns the result.

=cut

sub tcc_eval ($;@)
{
  my($code, @args) = @_;
  my $tcc = FFI::TinyCC->new;
  
  my $h = (caller(0))[10];
  if($h->{"FFI::TinyCC::Inline/options"})
  { $tcc->set_options($h->{"FFI::TinyCC::Inline/options"}) }

  $tcc->compile_string($code);
  my $meta = FFI::TinyCC::Parser->extract_function_metadata($code);
  my $func = $meta->{functions}->{main};
  croak "no main function" unless defined $func;
  my $sub = _generate_sub('main', $meta->{functions}->{main}, $tcc);
  $sub->(@args);
}

package
  FFI::TinyCC::Parser;

# this parser code stolen shamelessly
# from XS::TCC, which I strongly suspect
# was itself shamelessly "borrowed"
# from Inline::C::Parser::RegExp

# Copyright 2002 Brian Ingerson
# Copyright 2008, 2010-2012 Sisyphus
# Copyright 2013 Steffen Muellero

# This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

use strict;
use warnings;

# These regular expressions were derived from Regexp::Common v0.01.
my $RE_comment_C   = q{(?:(?:\/\*)(?:(?:(?!\*\/)[\s\S])*)(?:\*\/))};
my $RE_comment_Cpp = q{(?:\/\*(?:(?!\*\/)[\s\S])*\*\/|\/\/[^\n]*\n)};
my $RE_quoted      = (q{(?:(?:\")(?:[^\\\"]*(?:\\.[^\\\"]*)*)(?:\")}
                     .q{|(?:\')(?:[^\\\']*(?:\\.[^\\\']*)*)(?:\'))});
my $RE_balanced_brackets;
$RE_balanced_brackets =
  qr'(?:[{]((?:(?>[^{}]+)|(??{$RE_balanced_brackets}))*)[}])';
my $RE_balanced_parens;
$RE_balanced_parens =
  qr'(?:[(]((?:(?>[^()]+)|(??{$RE_balanced_parens}))*)[)])';


sub _normalize_type {
  # Normalize a type for lookup in a typemap.
  my($type) = @_;

  # Remove "extern".
  # But keep "static", "inline", "typedef", etc,
  #  to cause desirable typemap misses.
  $type =~ s/\bextern\b//g;

  # Whitespace: only single spaces, none leading or trailing.
  $type =~ s/\s+/ /g;
  $type =~ s/^\s//; $type =~ s/\s$//;

  # Adjacent "derivative characters" are not separated by whitespace,
  # but _are_ separated from the adjoining text.
  # [ Is really only * (and not ()[]) needed??? ]
  $type =~ s/\*\s\*/\*\*/g;
  $type =~ s/(?<=[^ \*])\*/ \*/g;

  return $type;
}

sub extract_function_metadata {
  my (undef, $code) = @_;

  my $results = {
    function_names => [],
    functions => {},
  };

  # First, we crush out anything potentially confusing.
  # The order of these _does_ matter.
  $code =~ s/$RE_comment_C/ /go;
  $code =~ s/$RE_comment_Cpp/ /go;
  $code =~ s/^\#.*(\\\n.*)*//mgo;
  #$code =~ s/$RE_quoted/\"\"/go; # Buggy, if included.
  $code =~ s/$RE_balanced_brackets/{ }/go;

  # The decision of what is an acceptable declaration was originally
  # derived from Inline::C::grammar.pm version 0.30 (Inline 0.43).

  my $re_plausible_place_to_begin_a_declaration = qr {
    # The beginning of a line, possibly indented.
    # (Accepting indentation allows for C code to be aligned with
    #  its surrounding perl, and for backwards compatibility with
    #  Inline 0.43).
    (?m: ^ ) \s*
  }xo;

  # Instead of using \s , we dont tolerate blank lines.
  # This matches user expectation better than allowing arbitrary
  # vertical whitespace.
  my $sp = qr{[ \t]|\n(?![ \t]*\n)};

  my $re_type = qr {(
    (?: \w+ $sp* )+? # words
    (?: \*  $sp* )*  # stars
  )}xo;

  my $re_identifier = qr{ (\w+) $sp* }xo;
  while( $code =~ m{
          $re_plausible_place_to_begin_a_declaration
          ( $re_type $re_identifier $RE_balanced_parens $sp* (\;|\{) )
         }xgo)
  {
    my($type, $identifier, $args, $what) = ($2,$3,$4,$5);
    $args = "" if $args =~ /^\s+$/;

    my $need_threading_context = 0;
    my $is_decl     = $what eq ';';
    my $function    = $identifier;
    my $return_type = _normalize_type($type);
    my @arguments   = split ',', $args;

    #goto RESYNC if $is_decl && !$self->{data}{AUTOWRAP};
    goto RESYNC if exists $results->{functions}{$function};
    #goto RESYNC if !defined $self->{data}{typeconv}{valid_rtypes}{$return_type};

    my(@arg_names,@arg_types);
    my $dummy_name = 'arg1';

    my $argno = 0;
    foreach my $arg (@arguments) {
      # recognize threading context passing as part of first arg
      if ($argno++ == 0 and $arg =~ s/^\s*pTHX_?\s*//) {
        $need_threading_context = 1;
        next if $arg !~ /\S/;
      }

      my $arg_no_space = $arg;
      $arg_no_space =~ s/\s+//g;

      # If $arg_no_space is 'void', there will be no identifier.
      if( my($type, $identifier) =
          $arg =~ /^\s*$re_type(?:$re_identifier)?\s*$/o )
      {
        my $arg_name = $identifier;
        my $arg_type = _normalize_type($type);

        if((!defined $arg_name) && ($arg_no_space ne 'void')) {
          goto RESYNC if !$is_decl;
          $arg_name = $dummy_name++;
        }
        #goto RESYNC if ((!defined
        #    $self->{data}{typeconv}{valid_types}{$arg_type}) && ($arg_no_space ne 'void'));

        # Push $arg_name onto @arg_names iff it's defined. Otherwise ($arg_no_space
        # was 'void'), push the empty string onto @arg_names (to avoid uninitialized
        # warnings emanating from C.pm).
        defined($arg_name) ? push(@arg_names,$arg_name)
                           : push(@arg_names, '');
        if($arg_name) {push(@arg_types,$arg_type)}
        else {push(@arg_types,'')} # $arg_no_space was 'void' - this push() avoids 'uninitialized' warnings from C.pm
      }
      elsif($arg =~ /^\s*\.\.\.\s*$/) {
        push(@arg_names,'...');
        push(@arg_types,'...');
      }
      else {
        goto RESYNC;
      }
    }

    # Commit.
    push @{$results->{function_names}}, $function;
    $results->{functions}{$function}{return_type}= $return_type;
    $results->{functions}{$function}{arg_names} = [@arg_names];
    $results->{functions}{$function}{arg_types} = [@arg_types];
    $results->{functions}{$function}{need_threading_context} = $need_threading_context if $need_threading_context;

    next;

RESYNC:  # Skip the rest of the current line, and continue.
    $code =~ /\G[^\n]*\n/gc;
  }

  return $results;
}

1;

=head1 SEE ALSO

=over 4

=item L<FFI::TinyCC>

=item L<C::Blocks>

=back

=head1 BUNDLED SOFTWARE

This package also comes with a parser that was shamelessly stolen from L<XS::TCC>,
which I strongly suspect was itself shamelessly "borrowed" from 
L<Inline::C::Parser::RegExp>

The license details for the parser are:

Copyright 2002 Brian Ingerson
Copyright 2008, 2010-2012 Sisyphus
Copyright 2013 Steffen Muellero

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut

